websockify/websocket.py

```diff
@@ -1,1109 +1,825 @@
-        # TODO: figure out a way to actually log this information without
-        #       calling `log` in the signal handlers
-        self.terminate()
-
-    def do_SIGTERM(self, sig, stack):
-        # TODO: figure out a way to actually log this information without
-        #       calling `log` in the signal handlers
-        self.terminate()
-
-    def top_new_client(self, startsock, address):
-        """ Do something with a WebSockets client connection. """
-        # handler process
-        client = None
-        try:
+        WebSocketWantWriteError can be raised if there is insufficient
+        space in the underlying socket for the close message.
+        """
+        self.shutdown(socket.SHUT_RDWR, code, reason)
+        self._close()
+
+    def _recv(self):
+        # Fetches more data from the socket to the buffer
+        assert self.socket is not None
+
+        while True:
             try:
-                client = self.do_handshake(startsock, address)
-            except self.EClose:
-                _, exc, _ = sys.exc_info()
-                # Connection was not a WebSockets connection
-                if exc.args[0]:
-                    self.msg("%s: %s" % (address[0], exc.args[0]))
-            except WebSocketServer.Terminate:
+                data = self.socket.recv(4096)
+            except (socket.error, OSError):
+                exc = sys.exc_info()[1]
+                if hasattr(exc, 'errno'):
+                    err = exc.errno
+                else:
+                    err = exc[0]
+
+                if err == errno.EWOULDBLOCK:
+                    raise WebSocketWantReadError
+
                 raise
-            except Exception:
-                _, exc, _ = sys.exc_info()
-                self.msg("handler exception: %s" % str(exc))
-                self.vmsg("exception", exc_info=True)
-        finally:
-
-            if client and client != startsock:
-                # Close the SSL wrapped socket
-                # Original socket closed by caller
-                client.close()
-
-    def get_log_fd(self):
-        """
-        Get file descriptors for the loggers.
-        They should not be closed when the process is forked.
-        """
-        descriptors = []
-        for handler in self.logger.parent.handlers:
-            if isinstance(handler, logging.FileHandler):
-                descriptors.append(handler.stream.fileno())
 
-        return descriptors
+            if len(data) == 0:
+                return False
 
-    def start_server(self):
-        """
-        Daemonize if requested. Listen for for connections. Run
-        do_handshake() method for each connection. If the connection
-        is a WebSockets client then call new_websocket_client() method (which must
-        be overridden) for each new client connection.
+            self._recv_buffer += data
+
+            # Support for SSLSocket like objects
+            if hasattr(self.socket, "pending"):
```

Comment:
That sounds like premature optimisation. :)
This is more readable IMO, and we have no reason to believe `hasattr` is slow.